File: Archive/AtCoder/AtCoder_Beginner_Contest/300/315/F.cpp
#include<bits/stdc++.h>
using namespace std;
int N;
double X[10000],Y[10000];
double dp[10000][29];
#define INF 2e8
int main(){
  cin>>N;
  for(int i=0;i<N;i++)cin>>X[i]>>Y[i];
  auto d=[&](int i,int j){
    return sqrt((X[j]-X[i])*(X[j]-X[i])+(Y[j]-Y[i])*(Y[j]-Y[i]));
  };
  fill(dp[0],dp[10000],INF); dp[0][0]=0;
  for(int i=1;i<N;i++)for(int j=0;j<29;j++)for(int k=0;k<=j;k++)
    dp[i][j]=min(dp[i][j],dp[i-1-(j-k)][k]+d(i,i-1-(j-k)));
  double a=INF; for(int i=0;i<29;i++)
    a=min(a,dp[N-1][i]+(i>0?1<<(i-1):0));
  cout<<fixed<<setprecision(20)<<a<<endl;
}

File: Archive/AtCoder/AtCoder_Beginner_Contest/300/315/F.nim
include atcoder/extra/header/chaemon_header
let
  N=nextInt()
  P=newSeqWith(N,(X:nextFloat(),Y:nextFloat()))
  Inf=2.0*10.0^4*10.0^4
  m=min(log2(Inf).int,N)
var dp=newSeqWith(N+1,newSeqWith(m+1,Inf))
dp[1][0]=0
proc d(a,b:int):float=
  return ((P[b-1].X-P[a-1].X)^2+(P[b-1].Y-P[a-1].Y)^2).sqrt
for i in 2..N:
  for j in 1..i-1:
    for k in 0..m:
      if k+i-j-1<=m:
        dp[i][k+i-j-1].min=dp[j][k]+d(j,i)
echo (block: collect(newSeq):
  for i in 0..m:
    dp[N][i]+(if i>0: 2.0^(i-1) else: 0.0)).min